src/content/reference/react-dom/components/form.md

```diff
@@ -0,0 +1,451 @@
+        <div key={index}>
+          {message.text}
+          {!!message.sending && <small> (Sending...)</small>}
+        </div>
+      ))}
+      <form action={formAction} id="message-form">
+        <input type="text" name="message" placeholder="Hello!" />
+        <button type="submit">Send</button>
+      </form>
+    </>
+  );
+}
+
+export default function App() {
+  const [messages, setMessages] = useState([
+    { text: "Hello there!", sending: false, key: 1 }
+  ]);
+  async function sendMessage(formData) {
+    const sentMessage = await deliverMessage(formData.get("message"));
+    setMessages([...messages, { text: sentMessage }]);
+  }
+  return <Thread messages={messages} sendMessage={sendMessage} />;
+}
+```
+
+```js actions.js
+export async function deliverMessage(message) {
+  await new Promise((res) => setTimeout(res, 1000));
+  return message;
+}
+```
+
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "18.3.0-canary-6db7f4209-20231021",
+    "react-dom": "18.3.0-canary-6db7f4209-20231021",
+    "react-scripts": "^5.0.0"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+
+</Sandpack>
+
+### Handling form submission errors {/*handling-form-submission-errors*/}
+
+In some cases the function called by a `<form>`'s `action` prop throw an error. You can handle these errors by wrapping `<form>` in an Error Boundary. If the function called by a `<form>`'s `action` prop throws an error, the fallback for the error boundary will be displayed.
+
+<Sandpack>
+
+```js App.js
+import { ErrorBoundary } from "react-error-boundary";
+import { action } from "./actions.js"
+
+export default function Search() {
+  return (
+    <ErrorBoundary fallback={<p>There was an error while submitting the form</p>}>
+      <form action={action}>
+        <input name="query" />
+        <button type="submit">Search</button>
+      </form>
+    </ErrorBoundary>
+
+  );
+}
+```
+
+```js actions.js hidden
+export async function search() {
+  return new Promise((_, reject) => reject())
+}
+```
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "18.3.0-canary-6db7f4209-20231021",
+    "react-dom": "18.3.0-canary-6db7f4209-20231021",
+    "react-scripts": "^5.0.0",
+    "react-error-boundary": "4.0.3"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+
+</Sandpack>
+
+To surface error messages to your users from a form action you can use the `useFormState` Hook. `useFormState` takes two parameters: an action and a inital state, and returns two values, a state varible and a form action. The form action is passed to the `action` prop of the form. The value returned by the action passed to `useFormState` is used to update the state varible. This can be used to return error message that can be displayed to a user. When use in conjunction with [Server Actions](/reference/react/use-server) and [Server Components](/reference/react/use-client), the error message is progressively enhanced.
```

Comment:
```suggestion
To surface error messages to your users from a form action you can use the `useFormState` Hook. `useFormState` takes two parameters: an action and a inital state, and returns two values, a state variable and a form action. The form action is passed to the `action` prop of the form. The value returned by the action passed to `useFormState` is used to update the state variable. This can be used to return error message that can be displayed to a user. When use in conjunction with [Server Actions](/reference/react/use-server) and [Server Components](/reference/react/use-client), the error message is progressively enhanced.
```

Reply:
I'm having a hard time following this..I also haven't read the `useFormState` docs yet so bear that in mind. 

I think first off the bit about "To surface error messages to your users from a form action", that doesn't feel clear to me why I need `useFormState`, couldn't I show an error message without the hook? 

Is there a mental model around `useFormState` we could share? I think the one I'm shaping in my mind reading this is that it allows you to receive a return value from your server action? But isn't that possible just from returning a value from the server function? But I guess you'd need a wrapper to read it.. so I guess that's what `useFormState` is for?

Reply:
Also, I think for this "`useFormState` takes two parameters: an action and a inital state", I'd suggestion saying function instead of action. Or maybe expanding on what action means here?

Reply:
The simple use case of useFormState, and probably what the great majority of uses will be, is to read the return value from a server action. The return value might be an error message but not of the sort that you'd want to throw an exception for — for example, if you can't add an item to your cart because it's sold out. That's an application logic thing, not a system-level failure.

The fully general use case, which explains the name of the hook, is to implement a reducer state machine that can be progressively enhanced. This is because useFormState remembers the previous return value from the server and passes it back to the server again each time the action is invoked.

The API is quite weird and difficult to explain so I'm not surprised that trying to compress it into a paragraph is hard to follow. 

Here's another attempt:

To surface error messages to your users from a form action you can use the `useFormState` Hook. This Hook creates a state variable that is set to the most recent value returned by the form action. It takes two parameters: an action and an initial state; and returns two values: the latest state and a wrapped version of the form action. You pass this wrapped action into your form. The action that you pass receives the previous state as its first argument. When use in conjunction with [Server Actions](/reference/react/use-server) and [Server Components](/reference/react/use-client), the error message is progressively enhanced.

Reply:
I'm not sure we need to fully explain the signature of useFormState here. Just a sentence about what it does / why to use it, and then the code sample and link to its own page can speak for themselves.

Reply:
gotcha, thanks for the feedback, this is pretty confusing. I broke the useFormState example out into its own section and rewrote the wording but kept the example the same.